packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_function_in_event_binding/index.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  AST,
+  ASTWithSource,
+  Call,
+  Chain,
+  Conditional,
+  ParsedEventType,
+  ParseSourceSpan,
+  PropertyRead,
+  SafeCall,
+  SafePropertyRead,
+  TmplAstBoundEvent,
+  TmplAstNode,
+} from '@angular/compiler';
+import ts from 'typescript';
+
+import {ErrorCode, ExtendedTemplateDiagnosticName} from '../../../../diagnostics';
+import {NgTemplateDiagnostic, SymbolKind} from '../../../api';
+import {TemplateCheckFactory, TemplateCheckWithVisitor, TemplateContext} from '../../api';
+
+/**
+ * Ensures that function in event bindings are called. For example, `<button (click)="myFunc"></button>`
+ * will not call `myFunc` when the button is clicked. Instead, it should be `<button (click)="myFunc()"></button>`.
+ * This is likely not the intent of the developer. Instead, the intent is likely to call `myFunc`.
+ */
+class UninvokedFunctionInEventBindingSpec extends TemplateCheckWithVisitor<ErrorCode.UNINVOKED_FUNCTION_IN_EVENT_BINDING> {
+  override code = ErrorCode.UNINVOKED_FUNCTION_IN_EVENT_BINDING as const;
+
+  override visitNode(
+    ctx: TemplateContext<ErrorCode.UNINVOKED_FUNCTION_IN_EVENT_BINDING>,
+    component: ts.ClassDeclaration,
+    node: TmplAstNode | AST,
+  ): NgTemplateDiagnostic<ErrorCode.UNINVOKED_FUNCTION_IN_EVENT_BINDING>[] {
+    // If the node is not a bound event, skip it.
+    if (!(node instanceof TmplAstBoundEvent)) return [];
+
+    // If the node is not a regular or animation event, skip it.
+    if (node.type !== ParsedEventType.Regular && node.type !== ParsedEventType.Animation) return [];
+
+    if (!(node.handler instanceof ASTWithSource)) return [];
+
+    const sourceExpressionText = node.handler.source || '';
+
+    // If the method is called, skip it.
+    if (node.handler.ast instanceof Call || node.handler.ast instanceof SafeCall) return [];
+
+    if (node.handler.ast instanceof Chain) {
+      // (click)="increment; decrement"
+      return node.handler.ast.expressions.flatMap((expression) =>
+        assertExpressionInvoked(expression, component, node, sourceExpressionText, ctx),
+      );
+    }
+
+    if (node.handler.ast instanceof Conditional) {
+      // (click)="true ? increment : decrement"
+      const {trueExp, falseExp} = node.handler.ast;
+      return [trueExp, falseExp].flatMap((expression) =>
+        assertExpressionInvoked(expression, component, node, sourceExpressionText, ctx),
+      );
+    }
+
+    // (click)="increment"
+    const symbol = ctx.templateTypeChecker.getSymbolOfNode(node.handler, component);
+    if (symbol !== null && symbol.kind === SymbolKind.Expression) {
+      if (symbol.tsType.getCallSignatures()?.length) {
+        return [generateDiagnostic(ctx, sourceExpressionText, node.sourceSpan)];
+      }
+    }
+
+    return [];
```

Comment:
You can avoid the duplicate logic here:

```suggestion
    return assertExpressionInvoked(node.handler.ast, component, node, sourceExpressionText, ctx)
```

Doing so also allows the above `if (node.handler.ast instanceof Call || node.handler.ast instanceof SafeCall) return [];` to be dropped, since that is also done in `assertExpressionInvoked`. You may even consider inlining all of `generateDiagnostic` into its single caller at that point.